src/Chapter3.hs

```diff
@@ -1006,6 +1190,122 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+class Fighter a where
+  getAttack :: a -> Attack
+  getHealth :: a -> Health
+  getDefense :: a -> Defense
+  getName :: a -> String
+  updateHealth :: a -> Health -> a
+  takeAction :: Fighter b => a -> b -> String
+  hit :: Fighter b => a -> b -> b
+  hit x y = takeDamage y $ getAttack x
+  heal :: a -> a
+  heal = id
+  castSpell :: a -> a
+  castSpell = id
+  takeDamage :: a -> Attack -> a
+  takeDamage fighter (MkAttack attack) =
+    let (MkDefense defense) = getDefense fighter
+        (MkHealth health) = getHealth fighter
+    in updateHealth fighter $
+        MkHealth $ health - max 0 (attack - defense)
+
+data Knight = Knight
+  { knightName :: String
+  , knightHealth :: Health
+  , knightAttack :: Attack
+  , knightDefense :: Defense
+  , knightActions :: [KnightAction]
+  } deriving Show
+
+instance Fighter Knight where
+  getName    = knightName
+  getAttack  = knightAttack
+  getHealth  = knightHealth
+  getDefense = knightDefense
+  updateHealth knight health = knight { knightHealth = health }
+  heal knight@Knight {..} =
+    let MkHealth h = knightHealth in knight { knightHealth = MkHealth $ h + 5 }
+  castSpell knight@Knight {..} =
+    let MkDefense d = knightDefense
+    in  knight { knightDefense = MkDefense $ d + 1 }
+  takeAction attacker'@Knight {..} defender = act nextAction attacker defender
+   where
+    (nextAction, actions) = getNextAction knightActions
+    attacker              = attacker' { knightActions = actions }
+
+data Monster = Monster
+  { monsterName :: String
+  , monsterHealth :: Health
+  , monsterAttack :: Attack
+  , monsterActions :: [MonsterAction]
+  } deriving Show
+
+
+instance Fighter Monster where
+  getName   = monsterName
+  getAttack = monsterAttack
+  getHealth = monsterHealth
+  getDefense _ = MkDefense 0
+  updateHealth monster health = monster { monsterHealth = health }
+  takeAction attacker'@Monster {..} defender = act nextAction attacker defender
+   where
+    (nextAction, actions) = getNextAction monsterActions
+    attacker              = attacker' { monsterActions = actions }
+
+class Action a where
+  act :: (Fighter f1, Fighter f2) => a -> f1 -> f2 -> String
+  getNextAction :: [a] -> (a, [a])
+  getNextAction (action:actions) = (action, actions++[action])
+  getNextAction [] = error "You must have an action to take"
```

Comment:
`error`s are usually discouraged for exceptions, and they should only be used when you can't prove the compiler that this will never happen. And many of such cases can be eliminated by making illegal states unrepresentable.

For instance, if you want to guarantee that the given list must be non-empty, you should use the `NonEmpty` list data type:

* http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List-NonEmpty.html

Also, the `getNextAction` doesn't use any of `Action` methods, and it actually doesn't require any constraints at all. It can be a simple top-level function :slightly_smiling_face:

Reply:
That's a really nice type. I like that it works so similar to a list, almost no difference in the refactor to use it. 

Glad to get rid of that `error` too :smile: